beacon-chain/blockchain/forkchoice/service.go

```diff
@@ -0,0 +1,460 @@
+	activeIndices, err := helpers.ActiveValidatorIndices(lastJustifiedState, lastJustifiedEpoch)
+	if err != nil {
+		return 0, errors.Wrap(err, "could not get active indices for last justified checkpoint")
+	}
+
+	wantedBlk, err := s.db.Block(s.ctx, bytesutil.ToBytes32(root))
+	if err != nil {
+		return 0, errors.Wrap(err, "could not get target block")
+	}
+
+	balances := uint64(0)
+	for _, i := range activeIndices {
+		if s.db.HasValidatorLatestVote(s.ctx, i) {
+			vote, err := s.db.ValidatorLatestVote(s.ctx, i)
+			if err != nil {
+				return 0, errors.Wrapf(err, "could not get validator %d's latest vote", i)
+			}
+
+			wantedRoot, err := s.ancestor(vote.Root, wantedBlk.Slot)
+			if err != nil {
+				return 0, errors.Wrapf(err, "could not get ancestor root for slot %d", wantedBlk.Slot)
+			}
+			if bytes.Equal(wantedRoot, root) {
+				balances += lastJustifiedState.Validators[i].EffectiveBalance
+			}
+		}
+	}
+	return balances, nil
+}
+
+// Head returns the head of the beacon chain.
+//
+// Spec pseudocode definition:
+//   def get_head(store: Store) -> Hash:
+//    # Execute the LMD-GHOST fork choice
+//    head = store.justified_checkpoint.root
+//    justified_slot = compute_start_slot_of_epoch(store.justified_checkpoint.epoch)
+//    while True:
+//        children = [
+//            root for root in store.blocks.keys()
+//            if store.blocks[root].parent_root == head and store.blocks[root].slot > justified_slot
+//        ]
+//        if len(children) == 0:
+//            return head
+//        # Sort by latest attesting balance with ties broken lexicographically
+//        head = max(children, key=lambda root: (get_latest_attesting_balance(store, root), root))
+func (s *Store) Head() ([]byte, error) {
+	head := s.justifiedCheckpt.Root
+
+	for {
+		startSlot := s.justifiedCheckpt.Epoch * params.BeaconConfig().SlotsPerEpoch
+		filter := filters.NewFilter().SetParentRoot(head).SetStartSlot(startSlot)
+		children, err := s.db.BlockRoots(s.ctx, filter)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not retrieve children info")
+		}
+
+		if len(children) == 0 {
+			return head, nil
+		}
+
+		// if a block has one child, then we don't have to lookup anything to
+		// know that this child will be the best child.
+		head = children[0]
+		if len(children) > 1 {
+			highest, err := s.latestAttestingBalance(head)
+			if err != nil {
+				return nil, errors.Wrap(err, "could not get latest balance")
+			}
+			for _, child := range children[1:] {
+				balance, err := s.latestAttestingBalance(child)
+				if err != nil {
+					return nil, errors.Wrap(err, "could not get latest balance")
+				}
+
+				if balance > highest {
+					highest = balance
+					head = child
+				}
+			}
+		}
+	}
+}
+
+// OnTick tracks the last unix time of when a block or an attestation arrives.
+//
+// Spec pseudocode definition:
+//   def on_tick(store: Store, time: uint64) -> None:
+//    store.time = time
+func (s *Store) OnTick(t uint64) {
```

Comment:
What is this used for?

Reply:
To update the time on when blockchain service last received block/attestation. It's called in `blockchain` service or any future service that needs to run fork choice. 

The time is used for the following checks:
```python
    # Blocks cannot be in the future. If they are, their consideration must be delayed until the are in the past.
    assert store.time >= pre_state.genesis_time + block.slot * SECONDS_PER_SLOT
```
```python
    # Blocks cannot be in the future. If they are, their consideration must be delayed until the are in the past.
    assert store.time >= pre_state.genesis_time + block.slot * SECONDS_PER_SLOT
```
```python
    # Blocks cannot be in the future. If they are, their consideration must be delayed until the are in the past.
    assert store.time >= pre_state.genesis_time + block.slot * SECONDS_PER_SLOT
```
```python
    # Attestations cannot be from future epochs. If they are, delay consideration until the epoch arrives
    assert store.time >= base_state.genesis_time + compute_start_slot_of_epoch(target.epoch) * SECONDS_PER_SLOT
```
```python
    # Attestations can only affect the fork choice of subsequent slots.
    # Delay consideration in the fork choice until their slot is in the past.
    assert store.time >= (attestation_slot + 1) * SECONDS_PER_SLOT
```